microsoft/vscode
# Fixes #32264 - Allow N instances of the same task

# Summary
* These changes allow N instances of the same task to be run simultaneously.
* The maximum number of allowable instances can be set in `RunOptions`.
* The instances of a task are visible in the QuickPick with an ` (i)` appended to the quick pick entry label to indicate that the task is an instance of an already running task. 
For example if 3 instances of task `foo` are running, the QuickPick would display:
```
foo
foo (1)
foo (2)
```
# Implementation notes
## Distinguishing between multiple instances of the same task
When a task is run in `terminalTaskSystem`, if another instance can be created, the task that the user selected is cloned and its `id` is modified to have form `<ORIGINAL_TASK_ID>|<COUNTER>`.
The counter gets incremented each time an instance of the same task is created. The cloned task is then run like any other task.
This makes it possible to check if a task is an instance of another task because all instances of a task will have the `<ORIGINAL_TASK_ID>` substring in common in their respective ids.
## Tracking multiple instances of the same task.
The `<ORIGINAL_TASK_ID>` substring shared by all instances serves as the key in a dictionary where the value is an `InstanceManager` object that stores information related to all instances of a task such as the number of currently active instances.

## How to test changes
Create a `tasks.json` and set the `instances` property. Here is an example configuration:
```json
{
    "version": "2.0.0",
    "tasks": [
        {
            "label": "echo",
            "type": "shell",
            "command": "sleep 20 && echo Hello",
            "runOptions": { "instances": 3}
        }
    ]
}
```
You can then execute the `echo` task, ensure that it never runs more than 3 instances of itself and that if a 4th instance is attempted, you should be prompted to either terminate or restart **the last instance** of `echo`.
